### ISSUE TITLE
Ensure the map can evolve over time (edit questions but keep a history, edit tasks and keep a history)

### ISSUE THREAD

**Issue by xehu** [Watts-Lab/task-mapping#345]

**Comment by xehu** [comment]
Current working version solves the task writeup unique ID problem:

[Comment copied from : https://github.com/Watts-Lab/multi-task-empirica/issues/65]

I added some code where --- and this requires a slight manual process --- we basically sub in any writeups that have to be edited as a NEW ROW. (Code + comments reproduced below, from commit [c9e3582f96c03efd8be7a11288ee79d6fe163074](https://github.com/Watts-Lab/task-mapping/compare/bf61070527d5...c9e3582f96c0).)

------------------
Prior to CSV generation: handle cases where we wrote new writeups but things already got rated
Basically, we want to generate new indices for tasks where we're now explicitly saving a new version for.
- If something hasn't been rated yet, we can go ahead and update the writeup. It's fine.
- If something is already rated, we need to generate a new entry (new row on the CSV) for it so that we can store it as a separate ID
- So this means that, in order to create a stable index:
     * we will manually set the ID to a known particular ID
     * the value of the ID will be N+1, where N is the number of tasks that are currently being actively rated in the pipeline
     * each time this happens again, we'll have to set that manual ID again, but this will allow that ID to remain consistent throughout the process.
```{r}
index_in_survey_1 <- 26
# Putting Food into Categories and Whac a Mole
new_tasks <- c("Putting food into categories-2", "Whac-A-Mole-2")
all_tasks <- c(df.survey_results$`task-name`[1:index_in_survey_1], new_tasks, df.survey_results$`task-name`[(index_in_survey_1+1):length(df.survey_results$`task-name`)])
```
------------------

I designed it this way so that we basically have a new row for every new "version;" the manual insertion guarantees that the row number is always unique. Then, I remove the previous version of the writeup from the randomization (but we are able to keep the data). From this point forward, we will collect labels for the updated writeups, which will have the same task name but a distinct ID in the system.

**Comment by xehu** [comment]
Tl;DR, if we need to update a task writeup, the process is:
1. **Create a duplicate (new) .md file,** so that we can save the new task writeup as a separate document. Name it `[Task Name]-#-writeup.md`. The # indicates the version number. The suffix needs to be `-writeup.md` in order to match all writeups (this suffix is automatically added to the writeups generated from the writing survey).
2. **Go to the pipeline, and manually add a line of code that creates a manual row for the new .md file.** Again, this has to be manual because we are basically circumventing the otherwise automated process that pulls the .md's from GitHub and automatically generates a CSV of HTML's. When doing this step, we also create a consistent index for the manual row so that we can track it.
3. **Run the code.** This generates the CSV of questions.
4. **Update the questions on Qualtrics using the CSV.** (This just involves adding the updated questions into the the Qualtrics Loop and Merge.)
5. **Remove the old versions of the questions from the Loop and Merge on Qualtrics.**

**Comment by xehu** [comment]
Note that we already have to do steps (3) and (4) in order to add any questions to Qualtrics anyway.

**Comment by markwhiting** [comment]
OK. Looking closer at this, I don't really like making new files because it means we're later going to have a cleanup that destroys history. I think just switching to index on a blob url instead of survey name would solve this, right? 

As in, any existing ratings are marked with the prior blob url, and moving forward we always just track that. 

Blob urls are permanent and pinpoint a historic version of a file. They can also be raw and could be used to generate the output you're using. Basically I'm suggesting something similar to what you have, but instead of making new files we just have a list of blobs that refer to versions of the task that have been rated (if we are careful with our data import, we could do this in a way that is data driven by qualtrics data, so we don't need to think about that list of blob urls explicitly)
 
Is there a good reason not to do this? Happy to jump on a call to talk through this design if any part seems unusual or annoying, though I think it means we get a better product that also maintains history beyond our current needs in a useful way.

**Comment by xehu** [comment]
Yes, I did run into an issue. The tl;dr is that I can’t figure out how to get blob indexing to work with loop and merge.

Basically, the way that things are currently happening is the following:

1. I read each written task in from the task writing survey
2. These task writeups then get transformed into .md files and saved in the directory/on GitHub
3. I read the .md files back into the script (we separately read them in from the directory because we may have manually made changes to the files outside of the survey). We read them in the same order that they were created in step 1, then turn them into a CSV, where they are converted to HTML strings
4. I copy/paste those questions into Qualtrics, **maintaining the same order in which they were read in** (the order is critical, as I explain in point 5). In fact, I keep all tasks basically in the same order throughout all steps to maintain consistency. This is why my current solution to maintain consistent index order is to “intervene” at step 3, where I insert new rows into the CSV at a static index for every “updated” task. This ensures each task writeup is always ordered the same. If I were to insert the rows for updated tasks at the end, for example, the order would be different each time as the total number of writeups increases.
5. Qualtrics Loop and Merge basically stores an underlying index; critically, **I can’t change this index / I am not sure how it can be modified and am not sure whether it can be a non-integer.** As far as I can tell, the index is determined under the hood by the order that the questions are entered into the survey. In fact, if you input 10 questions and then delete every other question, your indices end up being (1,3,5,7…. etc). So clearly the indices are a static under-the-hood entity associated with the order that the objects are inputted. Googling didn’t reveal a good way to set them to a custom value. 
Moreover, having a consistent index is essential because …
6. It identifies which questions belong to which columns. Otherwise, it’s impossible to process the 10,000-column data when Qualtrics exports the loop and merge results as a super sparse matrix. So I basically split out the loop index and use it to identify which question (task) it was. Because the indices match the tasks’ order in the CSV, I am able to retrieve the task names.

Keeping everything in the integer format also makes merging easier. if some tasks were indexed by integers and others were indexed by other things, it would get annoying to figure out what identifies what.

Thus, with all this context explained: I can’t use a blob to index because I am currently using the loop and merge index (note: I am not indexing by the survey name, but by an index generated by the ORDER in which task names appear in both the loop and merge and my representation of the tasks under the hood). This is probably the biggest sticking point — using another method breaks the way that I’m currently handling the question randomization, as well as the subsequent pipeline from loop and merge back into the task map in R.

A minor point that also makes blobs slightly harder is that I also don’t have access to the blob URL’s (I’m basically pulling in the .md files by traversing the file tree, and I’m unfamiliar with pulling blobs from the GitHub history.)

Hope this makes sense :)

**Comment by markwhiting** [comment]
> 5. Qualtrics Loop and Merge basically stores an underlying index; critically, **I can’t change this index / I am not sure how it can be modified and am not sure whether it can be a non-integer.** As far as I can tell, the index is determined under the hood by the order that the questions are entered into the survey. In fact, if you input 10 questions and then delete every other question, your indices end up being (1,3,5,7…. etc). So clearly the indices are a static under-the-hood entity associated with the order that the objects are inputted. Googling didn’t reveal a good way to set them to a custom value.
> 6. It identifies which questions belong to which columns. Otherwise, it’s impossible to process the 10,000-column data when Qualtrics exports the loop and merge results as a super sparse matrix. So I basically split out the loop index and use it to identify which question (task) it was. Because the indices match the tasks’ order in the CSV, I am able to retrieve the task names.

You don't need to deal with this because a raw qualtrics table puts the first column of the loop and merge in the header of the exported CSV, e.g., in this also massively sparse and huge matrix for commonsense, you can see that the statement with index 2 is "5 is a lot bigger than 1". I can then join on the raw text so I never care about the order of the qualtrics indices (which I completely agree are a total black box) 
<img width="671" alt="Screen Shot 2022-04-07 at 01 25 03" src="https://user-images.githubusercontent.com/673171/162126347-8b7d18ee-dc5b-4e62-b396-bcde648f609a.png">
 
In this case, if you make the first column of the loop and merge table the blob URL you will not only have a nice thing with no spaces in it, you will also know the blob URL from the data. And almost more importantly, you won't be relying on the order of elements to make sure things work out, which sounds frightening 😨 and risky 🔥. 

> A minor point that also makes blobs slightly harder is that I also don’t have access to the blob URL’s (I’m basically pulling in the .md files by traversing the file tree, and I’m unfamiliar with pulling blobs from the GitHub history.)

This is actually easier than it might sound. Here's how to get the blob URL for all the current versions (note the mention of `master` in the API URL).
```R
req <- httr::GET("https://api.github.com/repos/Watts-Lab/task-mapping/git/trees/master?recursive=1")
jsonlite::fromJSON(content(req, "text"))$tree %>% 
  filter(grepl("tasks/summaries/", path))
```
Moving forward, just get these, and if they change your system can automatically shift to include the appropriate new ones (blobs only change on commits that edit the referenced file). For the ones that have been edited, just replace `master` with the SHA of the commit before they were edited and the code will get you that set of URLs instead. 

You can of course get the contents of any of these blobs with something like: 
```R
file <- httr::GET("https://api.github.com/repos/Watts-Lab/task-mapping/git/blobs/770cc8101a99d4332bb17d91678e41e120c09979")
jsonlite::fromJSON(content(file, "text"))$content %>% jsonlite::base64_dec() %>% readBin("character")
```

The [Git Database API docs](https://docs.github.com/en/rest/reference/git#get-a-tree) talk about this particular call, and many others that could be useful, including something I just learnt about, [a markdown renderer](https://docs.github.com/en/rest/reference/markdown)!

**Comment by xehu** [comment]
With our switch to Glitch, we save all blobs --- and, in fact, we're able to compare old blob results to new blob results! It seems like we succeeded in this endeavor. Thank you, Mark! :) Let's close this issue.

**Comment by markwhiting** [comment]
Ok. I have one more step to do before I was going to close it but should be fine.
